CentOS 7 / RHEL 7: error compiling argument_parser

# Platform

  - SeqAn version: master branch (Also same error on 3.0.1)
  - Operating system: Linux cbrglogin2 3.10.0-1127.19.1.el7.x86_64 #1 SMP Tue Aug 25 17:23:54 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
  - Compiler: /opt/rh/devtoolset-9/root/usr/bin/g++
 - Compiler version - g++ (GCC) 9.3.1 20200408 (Red Hat 9.3.1-2)

## Description
Im relatively new to C++ so could entirely be my stupidity. When I try to compile my hello world I get the following error that seem to be related to the argument_parser because when I remove this header ref there are no errors. 

### How to repeat the problem

Code snippet:
```c++
#include <seqan3/core/debug_stream.hpp>
#include <seqan3/argument_parser/all.hpp>

int main()
{
    seqan3::debug_stream << "Hello world\n";
    return 0;
}
```
commands in build dir:
```
cmake ../source
make
```
Compile error:
```
[ 50%] Building CXX object CMakeFiles/hello_world.dir/hello_world.cpp.o
In file included from /Filers/home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/io/detail/misc.hpp:19,
                 from /Filers/home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/validators.hpp:29,
                 from /Filers/home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/detail/format_base.hpp:23,
                 from /Filers/home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/detail/format_help.hpp:20,
                 from /Filers/home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/argument_parser.hpp:25,
                 from /Filers/home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/all.hpp:15,
                 from /Filers/home/cgat/acribbs/Documents/tutorial/source/hello_world.cpp:3:
/Filers/home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/io/exception.hpp: In constructor ‘seqan3::io_error::io_error(const string&, const std::error_code&)’:
/Filers/home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/io/exception.hpp:59:72: error: no matching function for call to ‘std::ios_base::failure::failure(<brace-enclosed initializer list>)’
   59 |              std::error_code const & ec) : std::ios_base::failure{s, ec}
      |                                                                        ^
In file included from /opt/rh/devtoolset-9/root/usr/include/c++/9/ios:42,
                 from /opt/rh/devtoolset-9/root/usr/include/c++/9/ostream:38,
                 from /opt/rh/devtoolset-9/root/usr/include/c++/9/iostream:39,
                 from /Filers/home/cgat/acribbs/Documents/tutorial/source/hello_world.cpp:1:
/opt/rh/devtoolset-9/root/usr/include/c++/9/bits/ios_base.h:282:7: note: candidate: ‘std::ios_base::failure::failure(const string&)’
  282 |       failure(const string& __str) throw();
      |       ^~~~~~~
/opt/rh/devtoolset-9/root/usr/include/c++/9/bits/ios_base.h:282:7: note:   candidate expects 1 argument, 2 provided
/opt/rh/devtoolset-9/root/usr/include/c++/9/bits/ios_base.h:276:11: note: candidate: ‘std::ios_base::failure::failure(const std::ios_base::failure&)’
  276 |     class failure : public exception
      |           ^~~~~~~
/opt/rh/devtoolset-9/root/usr/include/c++/9/bits/ios_base.h:276:11: note:   candidate expects 1 argument, 2 provided
make[2]: *** [CMakeFiles/hello_world.dir/build.make:83: CMakeFiles/hello_world.dir/hello_world.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:96: CMakeFiles/hello_world.dir/all] Error 2
make: *** [Makefile:104: all] Error 2

```

Hi @marehr ,

The tutorials are great so far, they are very well written and explained!

Indeed SentOS is very similar to RedHat.

Thanks for your help

CmakeLists.txt:
`
cmake_minimum_required (VERSION 3.4)
project (seqan3_tutorial CXX)
 
find_package (SeqAn3 3.0.0 REQUIRED HINTS "${CMAKE_SOURCE_DIR}/../seqan3/build_system")
 
add_executable (hello_world hello_world.cpp)
 
target_link_libraries (hello_world seqan3::seqan3)
`

output of 'cmake ../source':
`[acribbs@cbrglogin2 build]$ cmake ../source 
-- The CXX compiler identification is GNU 7.3.1
-- Check for working CXX compiler: /opt/rh/devtoolset-7/root/usr/bin/c++
-- Check for working CXX compiler: /opt/rh/devtoolset-7/root/usr/bin/c++ - works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Finding SeqAn3 and checking requirements:
--   SeqAn3 include dir found:   /home/cgat/acribbs/Documents/tutorial/seqan3/include
--   Detected as running from a repository checkout…
--     …adding submodule include:  /home/cgat/acribbs/Documents/tutorial/seqan3/submodules/cereal/include
--     …adding submodule include:  /home/cgat/acribbs/Documents/tutorial/seqan3/submodules/range-v3/include
--     …adding submodule include:  /home/cgat/acribbs/Documents/tutorial/seqan3/submodules/sdsl-lite/include
--   C++ Standard-17 support:    via -std=c++17
--   C++ Concepts support:       via -fconcepts
--   C++ Filesystem header:      <experimental/filesystem>
--   C++ Filesystem library:     via -lstdc++fs
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
--   Thread support:             via -pthread
--   Required dependency:        Range-V3 found.
--   Required dependency:        SDSL found.
--   Optional dependency:        Cereal found.
--   Optional dependency:        Lemon not found.
--   Optional dependency:        ZLIB-1.2.7 found.
--   Optional dependency:        BZip2-1.0.6 found.
--   Optional dependency:        libexecinfo found.
--   SeqAn3 platform.hpp build:  passed.
-- Found SeqAn3: /home/cgat/acribbs/Documents/tutorial/seqan3/include (Required is at least version "3.0.0") 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/cgat/acribbs/Documents/tutorial/build
`
Output from make VERBOSE=1

`[acribbs@cbrglogin2 build]$ make VERBOSE=1
/Filers/package/cmake/3.17.2/bin/cmake -S/home/cgat/acribbs/Documents/tutorial/source -B/home/cgat/acribbs/Documents/tutorial/build --check-build-system CMakeFiles/Makefile.cmake 0
/Filers/package/cmake/3.17.2/bin/cmake -E cmake_progress_start /home/cgat/acribbs/Documents/tutorial/build/CMakeFiles /home/cgat/acribbs/Documents/tutorial/build/CMakeFiles/progress.marks
make  -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/Filers/home/cgat/acribbs/Documents/tutorial/build'
make  -f CMakeFiles/hello_world.dir/build.make CMakeFiles/hello_world.dir/depend
make[2]: Entering directory '/Filers/home/cgat/acribbs/Documents/tutorial/build'
cd /home/cgat/acribbs/Documents/tutorial/build && /Filers/package/cmake/3.17.2/bin/cmake -E cmake_depends "Unix Makefiles" /home/cgat/acribbs/Documents/tutorial/source /home/cgat/acribbs/Documents/tutorial/source /home/cgat/acribbs/Documents/tutorial/build /home/cgat/acribbs/Documents/tutorial/build /home/cgat/acribbs/Documents/tutorial/build/CMakeFiles/hello_world.dir/DependInfo.cmake --color=
Dependee "/home/cgat/acribbs/Documents/tutorial/build/CMakeFiles/hello_world.dir/DependInfo.cmake" is newer than depender "/home/cgat/acribbs/Documents/tutorial/build/CMakeFiles/hello_world.dir/depend.internal".
Dependee "/home/cgat/acribbs/Documents/tutorial/build/CMakeFiles/CMakeDirectoryInformation.cmake" is newer than depender "/home/cgat/acribbs/Documents/tutorial/build/CMakeFiles/hello_world.dir/depend.internal".
Scanning dependencies of target hello_world
make[2]: Leaving directory '/Filers/home/cgat/acribbs/Documents/tutorial/build'
make  -f CMakeFiles/hello_world.dir/build.make CMakeFiles/hello_world.dir/build
make[2]: Entering directory '/Filers/home/cgat/acribbs/Documents/tutorial/build'
[ 50%] Building CXX object CMakeFiles/hello_world.dir/hello_world.cpp.o
/opt/rh/devtoolset-7/root/usr/bin/c++  -DSEQAN3_HAS_BZIP2=1 -DSEQAN3_HAS_ZLIB=1 -I/home/cgat/acribbs/Documents/tutorial/seqan3/include -isystem /home/cgat/acribbs/Documents/tutorial/seqan3/submodules/sdsl-lite/include -isystem /home/cgat/acribbs/Documents/tutorial/seqan3/submodules/range-v3/include -isystem /home/cgat/acribbs/Documents/tutorial/seqan3/submodules/cereal/include  -std=c++17 -fconcepts -pthread -o CMakeFiles/hello_world.dir/hello_world.cpp.o -c /home/cgat/acribbs/Documents/tutorial/source/hello_world.cpp
/home/cgat/acribbs/Documents/tutorial/source/hello_world.cpp:3:10: fatal error: FQFeeder/src/FastxParser.hpp: No such file or directory
 #include <FQFeeder/src/FastxParser.hpp>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/hello_world.dir/build.make:83: CMakeFiles/hello_world.dir/hello_world.cpp.o] Error 1
make[2]: Leaving directory '/Filers/home/cgat/acribbs/Documents/tutorial/build'
make[1]: *** [CMakeFiles/Makefile2:96: CMakeFiles/hello_world.dir/all] Error 2
make[1]: Leaving directory '/Filers/home/cgat/acribbs/Documents/tutorial/build'
make: *** [Makefile:104: all] Error 2
[acribbs@cbrglogin2 build]$ cd ../source/
[acribbs@cbrglogin2 source]$ ls
CMakeLists.txt  CMakeLists.txt~  #hello_world.cpp#  hello_world.cpp  hello_world.cpp~
[acribbs@cbrglogin2 source]$ rm -rf CMakeLists.txt
CMakeLists.txt   CMakeLists.txt~  
[acribbs@cbrglogin2 source]$ rm -rf CMakeLists.txt~
[acribbs@cbrglogin2 source]$ rm -rf \#hello_world.cpp# 
[acribbs@cbrglogin2 source]$ rm -rf hello_world.cpp~
[acribbs@cbrglogin2 source]$ emacs hello_world.cpp 
[acribbs@cbrglogin2 source]$ ls
CMakeLists.txt  hello_world.cpp  hello_world.cpp~
[acribbs@cbrglogin2 source]$ emacs CMakeLists.txt 
[acribbs@cbrglogin2 source]$ cd ../build/
[acribbs@cbrglogin2 build]$ cmake ../source/
-- Finding SeqAn3 and checking requirements:
--   SeqAn3 include dir found:   /home/cgat/acribbs/Documents/tutorial/seqan3/include
--   Detected as running from a repository checkout…
--     …adding submodule include:  /home/cgat/acribbs/Documents/tutorial/seqan3/submodules/cereal/include
--     …adding submodule include:  /home/cgat/acribbs/Documents/tutorial/seqan3/submodules/range-v3/include
--     …adding submodule include:  /home/cgat/acribbs/Documents/tutorial/seqan3/submodules/sdsl-lite/include
--   C++ Standard-17 support:    via -std=c++17
--   C++ Concepts support:       via -fconcepts
--   C++ Filesystem header:      <experimental/filesystem>
--   C++ Filesystem library:     via -lstdc++fs
--   Thread support:             via -pthread
--   Required dependency:        Range-V3 found.
--   Required dependency:        SDSL found.
--   Optional dependency:        Cereal found.
--   Optional dependency:        Lemon not found.
--   Optional dependency:        ZLIB-1.2.7 found.
--   Optional dependency:        BZip2-1.0.6 found.
--   Optional dependency:        libexecinfo found.
--   SeqAn3 platform.hpp build:  passed.
-- Configuring done
-- Generating done
-- Build files have been written to: /home/cgat/acribbs/Documents/tutorial/build
[acribbs@cbrglogin2 build]$ make VERBOSE=1
/Filers/package/cmake/3.17.2/bin/cmake -S/home/cgat/acribbs/Documents/tutorial/source -B/home/cgat/acribbs/Documents/tutorial/build --check-build-system CMakeFiles/Makefile.cmake 0
/Filers/package/cmake/3.17.2/bin/cmake -E cmake_progress_start /home/cgat/acribbs/Documents/tutorial/build/CMakeFiles /home/cgat/acribbs/Documents/tutorial/build/CMakeFiles/progress.marks
make  -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/Filers/home/cgat/acribbs/Documents/tutorial/build'
make  -f CMakeFiles/hello_world.dir/build.make CMakeFiles/hello_world.dir/depend
make[2]: Entering directory '/Filers/home/cgat/acribbs/Documents/tutorial/build'
cd /home/cgat/acribbs/Documents/tutorial/build && /Filers/package/cmake/3.17.2/bin/cmake -E cmake_depends "Unix Makefiles" /home/cgat/acribbs/Documents/tutorial/source /home/cgat/acribbs/Documents/tutorial/source /home/cgat/acribbs/Documents/tutorial/build /home/cgat/acribbs/Documents/tutorial/build /home/cgat/acribbs/Documents/tutorial/build/CMakeFiles/hello_world.dir/DependInfo.cmake --color=
Dependee "/home/cgat/acribbs/Documents/tutorial/source/hello_world.cpp" is newer than depends file "/home/cgat/acribbs/Documents/tutorial/build/CMakeFiles/hello_world.dir/depend.internal".
Clearing dependencies in "/home/cgat/acribbs/Documents/tutorial/build/CMakeFiles/hello_world.dir/depend.make".
Scanning dependencies of target hello_world
make[2]: Leaving directory '/Filers/home/cgat/acribbs/Documents/tutorial/build'
make  -f CMakeFiles/hello_world.dir/build.make CMakeFiles/hello_world.dir/build
make[2]: Entering directory '/Filers/home/cgat/acribbs/Documents/tutorial/build'
[ 50%] Building CXX object CMakeFiles/hello_world.dir/hello_world.cpp.o
/opt/rh/devtoolset-7/root/usr/bin/c++  -DSEQAN3_HAS_BZIP2=1 -DSEQAN3_HAS_ZLIB=1 -I/home/cgat/acribbs/Documents/tutorial/seqan3/include -isystem /home/cgat/acribbs/Documents/tutorial/seqan3/submodules/sdsl-lite/include -isystem /home/cgat/acribbs/Documents/tutorial/seqan3/submodules/range-v3/include -isystem /home/cgat/acribbs/Documents/tutorial/seqan3/submodules/cereal/include  -std=c++17 -fconcepts -pthread -o CMakeFiles/hello_world.dir/hello_world.cpp.o -c /home/cgat/acribbs/Documents/tutorial/source/hello_world.cpp
In file included from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/io/detail/misc.hpp:19:0,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/validators.hpp:29,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/detail/format_base.hpp:23,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/detail/format_help.hpp:20,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/argument_parser.hpp:25,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/all.hpp:15,
                 from /home/cgat/acribbs/Documents/tutorial/source/hello_world.cpp:1:
/home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/io/exception.hpp: In constructor ‘seqan3::io_error::io_error(const string&, const std::error_code&)’:
/home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/io/exception.hpp:59:67: error: no matching function for call to ‘std::ios_base::failure::failure(<brace-enclosed initializer list>)’
              std::error_code const & ec) : std::ios_base::failure{}
                                                                   ^
In file included from /opt/rh/devtoolset-7/root/usr/include/c++/7/ios:42:0,
                 from /opt/rh/devtoolset-7/root/usr/include/c++/7/ostream:38,
                 from /opt/rh/devtoolset-7/root/usr/include/c++/7/iostream:39,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/argument_parser.hpp:16,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/all.hpp:15,
                 from /home/cgat/acribbs/Documents/tutorial/source/hello_world.cpp:1:
/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/ios_base.h:282:7: note: candidate: std::ios_base::failure::failure(const string&)
       failure(const string& __str) throw();
       ^~~~~~~
/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/ios_base.h:282:7: note:   candidate expects 1 argument, 0 provided
/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/ios_base.h:276:11: note: candidate: std::ios_base::failure::failure(const std::ios_base::failure&)
     class failure : public exception
           ^~~~~~~
/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/ios_base.h:276:11: note:   candidate expects 1 argument, 0 provided
make[2]: *** [CMakeFiles/hello_world.dir/build.make:83: CMakeFiles/hello_world.dir/hello_world.cpp.o] Error 1
make[2]: Leaving directory '/Filers/home/cgat/acribbs/Documents/tutorial/build'
make[1]: *** [CMakeFiles/Makefile2:96: CMakeFiles/hello_world.dir/all] Error 2
make[1]: Leaving directory '/Filers/home/cgat/acribbs/Documents/tutorial/build'
make: *** [Makefile:104: all] Error 2
`
Output  of /opt/rh/devtoolset-9/root/usr/bin/g++ -v:

`[acribbs@cbrglogin2 build]$ /opt/rh/devtoolset-9/root/usr/bin/g++ -v
Using built-in specs.
COLLECT_GCC=/opt/rh/devtoolset-9/root/usr/bin/g++
COLLECT_LTO_WRAPPER=/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/lto-wrapper
Target: x86_64-redhat-linux
Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,fortran,lto --prefix=/opt/rh/devtoolset-9/root/usr --mandir=/opt/rh/devtoolset-9/root/usr/share/man --infodir=/opt/rh/devtoolset-9/root/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --with-linker-hash-style=gnu --with-default-libstdcxx-abi=gcc4-compatible --enable-plugin --enable-initfini-array --with-isl=/builddir/build/BUILD/gcc-9.3.1-20200408/obj-x86_64-redhat-linux/isl-install --disable-libmpx --enable-gnu-indirect-function --with-tune=generic --with-arch_32=x86-64 --build=x86_64-redhat-linux
Thread model: posix
gcc version 9.3.1 20200408 (Red Hat 9.3.1-2) (GCC) 
`

Same error im afraid with 
`cmake -DCMAKE_CXX_FLAGS="-D_GLIBCXX_USE_CXX11_ABI=1" ../source/`

```
[acribbs@cbrglogin2 source]$ make
Scanning dependencies of target hello_world
[ 50%] Building CXX object CMakeFiles/hello_world.dir/hello_world.cpp.o
In file included from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/io/detail/misc.hpp:19:0,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/validators.hpp:29,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/detail/format_base.hpp:23,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/detail/format_help.hpp:20,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/argument_parser.hpp:25,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/all.hpp:15,
                 from /home/cgat/acribbs/Documents/tutorial/source/hello_world.cpp:1:
/home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/io/exception.hpp: In constructor ‘seqan3::io_error::io_error(const string&, const std::error_code&)’:
/home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/io/exception.hpp:59:67: error: no matching function for call to ‘std::ios_base::failure::failure(<brace-enclosed initializer list>)’
              std::error_code const & ec) : std::ios_base::failure{}
                                                                   ^
In file included from /opt/rh/devtoolset-7/root/usr/include/c++/7/ios:42:0,
                 from /opt/rh/devtoolset-7/root/usr/include/c++/7/ostream:38,
                 from /opt/rh/devtoolset-7/root/usr/include/c++/7/iostream:39,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/argument_parser.hpp:16,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/all.hpp:15,
                 from /home/cgat/acribbs/Documents/tutorial/source/hello_world.cpp:1:
/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/ios_base.h:282:7: note: candidate: std::ios_base::failure::failure(const string&)
       failure(const string& __str) throw();
       ^~~~~~~
/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/ios_base.h:282:7: note:   candidate expects 1 argument, 0 provided
/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/ios_base.h:276:11: note: candidate: std::ios_base::failure::failure(const std::ios_base::failure&)
     class failure : public exception
           ^~~~~~~
/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/ios_base.h:276:11: note:   candidate expects 1 argument, 0 provided
make[2]: *** [CMakeFiles/hello_world.dir/build.make:83: CMakeFiles/hello_world.dir/hello_world.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:96: CMakeFiles/hello_world.dir/all] Error 2
make: *** [Makefile:104: all] Error 2
```

```
(base) [acribbs@cbrglogin2 build]$ make VERBOSE=1
/Filers/package/cmake/3.17.2/bin/cmake -S/home/cgat/acribbs/Documents/tutorial/source -B/home/cgat/acribbs/Documents/tutorial/build --check-build-system CMakeFiles/Makefile.cmake 0
/Filers/package/cmake/3.17.2/bin/cmake -E cmake_progress_start /home/cgat/acribbs/Documents/tutorial/build/CMakeFiles /home/cgat/acribbs/Documents/tutorial/build/CMakeFiles/progress.marks
make  -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/Filers/home/cgat/acribbs/Documents/tutorial/build'
make  -f CMakeFiles/hello_world.dir/build.make CMakeFiles/hello_world.dir/depend
make[2]: Entering directory '/Filers/home/cgat/acribbs/Documents/tutorial/build'
cd /home/cgat/acribbs/Documents/tutorial/build && /Filers/package/cmake/3.17.2/bin/cmake -E cmake_depends "Unix Makefiles" /home/cgat/acribbs/Documents/tutorial/source /home/cgat/acribbs/Documents/tutorial/source /home/cgat/acribbs/Documents/tutorial/build /home/cgat/acribbs/Documents/tutorial/build /home/cgat/acribbs/Documents/tutorial/build/CMakeFiles/hello_world.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/Filers/home/cgat/acribbs/Documents/tutorial/build'
make  -f CMakeFiles/hello_world.dir/build.make CMakeFiles/hello_world.dir/build
make[2]: Entering directory '/Filers/home/cgat/acribbs/Documents/tutorial/build'
[ 50%] Building CXX object CMakeFiles/hello_world.dir/hello_world.cpp.o
/opt/rh/devtoolset-7/root/usr/bin/c++  -DSEQAN3_HAS_BZIP2=1 -DSEQAN3_HAS_ZLIB=1 -I/home/cgat/acribbs/Documents/tutorial/seqan3/include -isystem /home/cgat/acribbs/Documents/tutorial/seqan3/submodules/sdsl-lite/include -isystem /home/cgat/acribbs/Documents/tutorial/seqan3/submodules/range-v3/include -isystem /home/cgat/acribbs/Documents/tutorial/seqan3/submodules/cereal/include  -std=c++17 -fconcepts -pthread -o CMakeFiles/hello_world.dir/hello_world.cpp.o -c /home/cgat/acribbs/Documents/tutorial/source/hello_world.cpp
In file included from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/io/detail/misc.hpp:19:0,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/validators.hpp:29,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/detail/format_base.hpp:23,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/detail/format_help.hpp:20,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/argument_parser.hpp:25,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/all.hpp:15,
                 from /home/cgat/acribbs/Documents/tutorial/source/hello_world.cpp:1:
/home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/io/exception.hpp: In constructor ‘seqan3::io_error::io_error(const string&, const std::error_code&)’:
/home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/io/exception.hpp:59:67: error: no matching function for call to ‘std::ios_base::failure::failure(<brace-enclosed initializer list>)’
              std::error_code const & ec) : std::ios_base::failure{}
                                                                   ^
In file included from /opt/rh/devtoolset-7/root/usr/include/c++/7/ios:42:0,
                 from /opt/rh/devtoolset-7/root/usr/include/c++/7/ostream:38,
                 from /opt/rh/devtoolset-7/root/usr/include/c++/7/iostream:39,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/argument_parser.hpp:16,
                 from /home/cgat/acribbs/Documents/tutorial/seqan3/include/seqan3/argument_parser/all.hpp:15,
                 from /home/cgat/acribbs/Documents/tutorial/source/hello_world.cpp:1:
/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/ios_base.h:282:7: note: candidate: std::ios_base::failure::failure(const string&)
       failure(const string& __str) throw();
       ^~~~~~~
/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/ios_base.h:282:7: note:   candidate expects 1 argument, 0 provided
/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/ios_base.h:276:11: note: candidate: std::ios_base::failure::failure(const std::ios_base::failure&)
     class failure : public exception
           ^~~~~~~
/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/ios_base.h:276:11: note:   candidate expects 1 argument, 0 provided
make[2]: *** [CMakeFiles/hello_world.dir/build.make:83: CMakeFiles/hello_world.dir/hello_world.cpp.o] Error 1
make[2]: Leaving directory '/Filers/home/cgat/acribbs/Documents/tutorial/build'
make[1]: *** [CMakeFiles/Makefile2:96: CMakeFiles/hello_world.dir/all] Error 2
make[1]: Leaving directory '/Filers/home/cgat/acribbs/Documents/tutorial/build'
make: *** [Makefile:104: all] Error 2
```

Hello,
 
I tried to reproduce your problem.
I compiled g++-9 with the `--with-default-libstdcxx-abi=gcc4-compatible` flag and get the same error as you.

<details><summary>Click me</summary>

```console
cmake ../seqan3/test/snippet/ -DCMAKE_CXX_COMPILER=/dev/shm/seiler/issue_2244/gcc-9/bin/g++

...

make VERBOSE=1 issue_2244_snippet

...

[100%] Building CXX object CMakeFiles/issue_2244_snippet.dir/argument_parser/issue_2244.cpp.o
/usr/bin/ccache /dev/shm/seiler/issue_2244/gcc-9/bin/g++  -DSEQAN3_HAS_BZIP2=1 -DSEQAN3_HAS_ZLIB=1 -I/dev/shm/seiler/issue_2244/issue_test/vendor/googletest/googletest/include -I/dev/shm/seiler/issue_2244/seqan3/test/include -I/dev/shm/seiler/issue_2244/seqan3/include -isystem /dev/shm/seiler/issue_2244/seqan3/submodules/sdsl-lite/include -isystem /dev/shm/seiler/issue_2244/seqan3/submodules/range-v3/include -isystem /dev/shm/seiler/issue_2244/seqan3/submodules/cereal/include  -Wno-unused -pedantic -Wall -Wextra -Werror -std=c++17 -fconcepts -pthread -o CMakeFiles/issue_2244_snippet.dir/argument_parser/issue_2244.cpp.o -c /dev/shm/seiler/issue_2244/seqan3/test/snippet/argument_parser/issue_2244.cpp
In file included from /dev/shm/seiler/issue_2244/seqan3/include/seqan3/io/detail/misc.hpp:19,
                 from /dev/shm/seiler/issue_2244/seqan3/include/seqan3/argument_parser/validators.hpp:29,
                 from /dev/shm/seiler/issue_2244/seqan3/include/seqan3/argument_parser/detail/format_base.hpp:23,
                 from /dev/shm/seiler/issue_2244/seqan3/include/seqan3/argument_parser/detail/format_help.hpp:20,
                 from /dev/shm/seiler/issue_2244/seqan3/include/seqan3/argument_parser/argument_parser.hpp:25,
                 from /dev/shm/seiler/issue_2244/seqan3/include/seqan3/argument_parser/all.hpp:15,
                 from /dev/shm/seiler/issue_2244/seqan3/test/snippet/argument_parser/issue_2244.cpp:2:
/dev/shm/seiler/issue_2244/seqan3/include/seqan3/io/exception.hpp: In constructor ‘seqan3::io_error::io_error(const string&, const std::error_code&)’:
/dev/shm/seiler/issue_2244/seqan3/include/seqan3/io/exception.hpp:59:72: error: no matching function for call to ‘std::ios_base::failure::failure(<brace-enclosed initializer list>)’                                                                                                                                           59 |              std::error_code const & ec) : std::ios_base::failure{s, ec}
      |                                                                        ^
In file included from /dev/shm/seiler/issue_2244/gcc-9/include/c++/9.3.1/ios:42,
                 from /dev/shm/seiler/issue_2244/gcc-9/include/c++/9.3.1/ostream:38,
                 from /dev/shm/seiler/issue_2244/gcc-9/include/c++/9.3.1/iostream:39,
                 from /dev/shm/seiler/issue_2244/seqan3/include/seqan3/core/debug_stream.hpp:15,
                 from /dev/shm/seiler/issue_2244/seqan3/test/snippet/argument_parser/issue_2244.cpp:1:
/dev/shm/seiler/issue_2244/gcc-9/include/c++/9.3.1/bits/ios_base.h:282:7: note: candidate: ‘std::ios_base::failure::failure(const string&)’
  282 |       failure(const string& __str) throw();
      |       ^~~~~~~
/dev/shm/seiler/issue_2244/gcc-9/include/c++/9.3.1/bits/ios_base.h:282:7: note:   candidate expects 1 argument, 2 provided
/dev/shm/seiler/issue_2244/gcc-9/include/c++/9.3.1/bits/ios_base.h:276:11: note: candidate: ‘std::ios_base::failure::failure(const std::ios_base::failure&)’
  276 |     class failure : public exception
      |           ^~~~~~~
/dev/shm/seiler/issue_2244/gcc-9/include/c++/9.3.1/bits/ios_base.h:276:11: note:   candidate expects 1 argument, 2 provided
make[3]: *** [CMakeFiles/issue_2244_snippet.dir/build.make:63: CMakeFiles/issue_2244_snippet.dir/argument_parser/issue_2244.cpp.o] Error 1
make[3]: Leaving directory '/dev/shm/seiler/issue_2244/issue_test'
make[2]: *** [CMakeFiles/Makefile2:13051: CMakeFiles/issue_2244_snippet.dir/all] Error 2
make[2]: Leaving directory '/dev/shm/seiler/issue_2244/issue_test'
make[1]: *** [CMakeFiles/Makefile2:13058: CMakeFiles/issue_2244_snippet.dir/rule] Error 2
make[1]: Leaving directory '/dev/shm/seiler/issue_2244/issue_test'
make: *** [Makefile:5914: issue_2244_snippet] Error 2
```

</details>

However,  using the `-DCMAKE_CXX_FLAGS="-D_GLIBCXX_USE_CXX11_ABI=1"` solved the issue for me.

<details><summary>Click me</summary>

```console
cmake ../seqan3/test/snippet/ -DCMAKE_CXX_FLAGS="-D_GLIBCXX_USE_CXX11_ABI=1" -DCMAKE_CXX_COMPILER=/dev/shm/seiler/issue_2244/gcc-9/bin/g++

...

make VERBOSE=1 issue_2244_snippet

...

Scanning dependencies of target issue_2244_snippet
make[3]: Leaving directory '/dev/shm/seiler/issue_2244/issue_test2'
make -f CMakeFiles/issue_2244_snippet.dir/build.make CMakeFiles/issue_2244_snippet.dir/build
make[3]: Entering directory '/dev/shm/seiler/issue_2244/issue_test2'
[100%] Building CXX object CMakeFiles/issue_2244_snippet.dir/argument_parser/issue_2244.cpp.o
/usr/bin/ccache /dev/shm/seiler/issue_2244/gcc-9/bin/g++  -DSEQAN3_HAS_BZIP2=1 -DSEQAN3_HAS_ZLIB=1 -I/dev/shm/seiler/issue_2244/issue_test2/vendor/googletest/googletest/include -I/dev/shm/seiler/issue_2244/seqan3/test/include -I/dev/shm/seiler/issue_2244/seqan3/include -isystem /dev/shm/seiler/issue_2244/seqan3/submodules/sdsl-lite/include -isystem /dev/shm/seiler/issue_2244/seqan3/submodules/range-v3/include -isystem /dev/shm/seiler/issue_2244/seqan3/submodules/cereal/include  -D_GLIBCXX_USE_CXX11_ABI=1   -Wno-unused -pedantic -Wall -Wextra -Werror -std=c++17 -fconcepts -pthread -o CMakeFiles/issue_2244_snippet.dir/argument_parser/issue_2244.cpp.o -c /dev/shm/seiler/issue_2244/seqan3/test/snippet/argument_parser/issue_2244.cpp
[100%] Linking CXX executable argument_parser/issue_2244_snippet
/usr/bin/cmake -E cmake_link_script CMakeFiles/issue_2244_snippet.dir/link.txt --verbose=1
/dev/shm/seiler/issue_2244/gcc-9/bin/g++  -D_GLIBCXX_USE_CXX11_ABI=1   CMakeFiles/issue_2244_snippet.dir/argument_parser/issue_2244.cpp.o  -o argument_parser/issue_2244_snippet  -pthread /usr/lib/powerpc64le-linux-gnu/libz.so /usr/lib/powerpc64le-linux-gnu/libbz2.so -lrt lib/libgtest_main.a lib/libgtest.a -lpthread make[3]: Leaving directory '/dev/shm/seiler/issue_2244/issue_test2'
[100%] Built target issue_2244_snippet
make[2]: Leaving directory '/dev/shm/seiler/issue_2244/issue_test2'
/usr/bin/cmake -E cmake_progress_start /dev/shm/seiler/issue_2244/issue_test2/CMakeFiles 0
make[1]: Leaving directory '/dev/shm/seiler/issue_2244/issue_test2'
seiler@flughafenberlinbrandenburgwillybrandt:/dev/shm/seiler/issue_22
```
</details>

Your `make VERBOSE=1` output looks like the `-DCMAKE_CXX_FLAGS="-D_GLIBCXX_USE_CXX11_ABI=1" wasn't incorporated.
Can you retry, or try to run cmake again in an empty folder?

--- a/include/seqan3/core/platform.hpp
+++ b/include/seqan3/core/platform.hpp
@@ -293,6 +293,24 @@
 #   endif
 #endif
 
+/*!\brief This is needed to support CentOS 7 or RHEL 7; Newer CentOS's include a more modern default-gcc version making
+ *        this macro obsolete.
+ *
+ * In GCC 5 there was a bigger ABI change and modern systems compile with dual ABI, but some enterprise systems (those
+ * where gcc 4 is the standard compiler) don't support dual ABI. This has the effect that even community builds of gcc
+ * are build with --disable-libstdcxx-dual-abi. Only building the compiler yourself would solve this problem.
+ *
+ * \see https://github.com/seqan/seqan3/issues/2244
+ * \see https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_dual_abi.html
+ */
+#ifndef SEQAN3_WORKAROUND_GCC_NO_CXX11_ABI
+#   if defined(_GLIBCXX_USE_CXX11_ABI) && _GLIBCXX_USE_CXX11_ABI == 0
+#       define SEQAN3_WORKAROUND_GCC_NO_CXX11_ABI 1
+#   else
+#       define SEQAN3_WORKAROUND_GCC_NO_CXX11_ABI 0
+#   endif
+#endif
+
 //!\brief Workaround to access the static id of the configuration elements inside of the concept definition
 //!       (fixed in gcc11).
 #ifndef SEQAN3_WORKAROUND_GCC_PIPEABLE_CONFIG_CONCEPT

--- a/include/seqan3/io/alignment_file/format_bam.hpp
+++ b/include/seqan3/io/alignment_file/format_bam.hpp
@@ -64,12 +64,13 @@ public:
     /*!\name Constructors, destructor and assignment
      * \{
      */
-    format_bam() noexcept = default; //!< Defaulted.
-    format_bam(format_bam const &) noexcept = default; //!< Defaulted.
-    format_bam & operator=(format_bam const &) noexcept = default; //!< Defaulted.
-    format_bam(format_bam &&) noexcept = default; //!< Defaulted.
-    format_bam & operator=(format_bam &&) noexcept = default; //!< Defaulted.
-    ~format_bam() noexcept = default; //!< Defaulted.
+    // string_buffer is of type std::string and has some problems with pre-C++11 ABI
+    format_bam() = default; //!< Defaulted.
+    format_bam(format_bam const &) = default; //!< Defaulted.
+    format_bam & operator=(format_bam const &) = default; //!< Defaulted.
+    format_bam(format_bam &&) = default; //!< Defaulted.
+    format_bam & operator=(format_bam &&) = default; //!< Defaulted.
+    ~format_bam() = default; //!< Defaulted.
     //!\}
 
     //!\brief The valid file extensions for this format; note that you can modify this value.

--- a/include/seqan3/io/alignment_file/format_sam_base.hpp
+++ b/include/seqan3/io/alignment_file/format_sam_base.hpp
@@ -66,12 +66,12 @@ protected:
     /*!\name Constructors, destructor and assignment
      * \{
      */
-    format_sam_base() noexcept = default; //!< Defaulted.
-    format_sam_base(format_sam_base const &) noexcept = default; //!< Defaulted.
-    format_sam_base & operator=(format_sam_base const &) noexcept = default; //!< Defaulted.
-    format_sam_base(format_sam_base &&) noexcept = default; //!< Defaulted.
-    format_sam_base & operator=(format_sam_base &&) noexcept = default; //!< Defaulted.
-    ~format_sam_base() noexcept = default; //!< Defaulted.
+    format_sam_base() = default; //!< Defaulted.
+    format_sam_base(format_sam_base const &) = default; //!< Defaulted.
+    format_sam_base & operator=(format_sam_base const &) = default; //!< Defaulted.
+    format_sam_base(format_sam_base &&) = default; //!< Defaulted.
+    format_sam_base & operator=(format_sam_base &&) = default; //!< Defaulted.
+    ~format_sam_base() = default; //!< Defaulted.
     //!\}
 
     //!\brief The format version string.

--- a/include/seqan3/io/exception.hpp
+++ b/include/seqan3/io/exception.hpp
@@ -54,14 +54,17 @@ struct parse_error : std::runtime_error
 //!\brief Thrown if there is an io error in low level io operations such as in std::basic_streambuf operations.
 struct io_error : std::ios_base::failure
 {
+#if SEQAN3_WORKAROUND_GCC_NO_CXX11_ABI
+    // std::ios_base::failure is missing the std::error_code constructor in pre-C++11 ABI
+    // see https://en.cppreference.com/w/cpp/io/ios_base/failure
+    using base_t = std::ios_base::failure;
+    using base_t::base_t;
+#else // ^^^ workaround / no workaround vvv
     //!\brief Constructor that forwards the exception string.
-    io_error(std::string const & s,
-             std::error_code const & ec) : std::ios_base::failure{s, ec}
+    explicit io_error(std::string const & s, std::error_code const & ec = std::io_errc::stream)
+        : std::ios_base::failure{s, ec}
     {}
-
-    //!\brief Constructor that forwards the exception string.
-    explicit io_error(std::string const & s) : io_error{s, std::error_code{std::io_errc::stream}}
-    {}
+#endif // SEQAN3_WORKAROUND_GCC_NO_CXX11_ABI
 };
 
 // ----------------------------------------------------------------------------

--- a/include/seqan3/range/container/concept.hpp
+++ b/include/seqan3/range/container/concept.hpp
@@ -162,13 +162,26 @@ SEQAN3_CONCEPT sequence_container = requires (type val, type val2, type const cv
     // TODO: how do you model this?
     // SEQAN3_RETURN_TYPE_CONSTRAINT(val.emplace(typename type::const_iterator{}, ?),
     //                               std::same_as, typename type::iterator);
-    SEQAN3_RETURN_TYPE_CONSTRAINT(val.insert(val.cbegin(), val2.front()), std::same_as, typename type::iterator);
-    SEQAN3_RETURN_TYPE_CONSTRAINT(val.insert(val.cbegin(), typename type::value_type{}),
+#if SEQAN3_WORKAROUND_GCC_NO_CXX11_ABI
+    SEQAN3_RETURN_TYPE_CONSTRAINT(val.insert(val.begin(), val2.front()), std::same_as, typename type::iterator);
+    SEQAN3_RETURN_TYPE_CONSTRAINT(val.insert(val.begin(), typename type::value_type{}),
                                   std::same_as, typename type::iterator);
-    SEQAN3_RETURN_TYPE_CONSTRAINT(val.insert(val.cbegin(), typename type::size_type{}, typename type::value_type{}),
-                                  std::same_as, typename type::iterator);
-    SEQAN3_RETURN_TYPE_CONSTRAINT(val.insert(val.cbegin(), val2.begin(), val2.end()),
-                                  std::same_as, typename type::iterator);
+
+    // std::string is missing the const_iterator versions for insert in pre-C++11 ABI
+    requires detail::is_basic_string_v<type> || requires(type val, type val2)
+#else // ^^^ workaround / no workaround vvv
+    requires requires(type val, type val2)
+#endif // SEQAN3_WORKAROUND_GCC_NO_CXX11_ABI
+    {
+        SEQAN3_RETURN_TYPE_CONSTRAINT(val.insert(val.cbegin(), val2.front()), std::same_as, typename type::iterator);
+        SEQAN3_RETURN_TYPE_CONSTRAINT(val.insert(val.cbegin(), typename type::value_type{}),
+                                      std::same_as, typename type::iterator);
+        SEQAN3_RETURN_TYPE_CONSTRAINT(val.insert(val.cbegin(), typename type::size_type{}, typename type::value_type{}),
+                                      std::same_as, typename type::iterator);
+        SEQAN3_RETURN_TYPE_CONSTRAINT(val.insert(val.cbegin(), val2.begin(), val2.end()),
+                                      std::same_as, typename type::iterator);
+    };
+
     requires detail::is_basic_string_v<type> || requires(type val)
     {
         // TODO this function is not defined on strings (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=83328)
@@ -176,8 +189,16 @@ SEQAN3_CONCEPT sequence_container = requires (type val, type val2, type const cv
                                       std::same_as, typename type::iterator);
     };
 
-    SEQAN3_RETURN_TYPE_CONSTRAINT(val.erase(val.cbegin()), std::same_as, typename type::iterator);
-    SEQAN3_RETURN_TYPE_CONSTRAINT(val.erase(val.cbegin(), val.cend()), std::same_as, typename type::iterator);
+#if SEQAN3_WORKAROUND_GCC_NO_CXX11_ABI
+    // std::string is missing the const_iterator versions for erase in pre-C++11 ABI
+    requires detail::is_basic_string_v<type> || requires(type val)
+#else // ^^^ workaround / no workaround vvv
+    requires requires(type val)
+#endif // SEQAN3_WORKAROUND_GCC_NO_CXX11_ABI
+    {
+        SEQAN3_RETURN_TYPE_CONSTRAINT(val.erase(val.cbegin()), std::same_as, typename type::iterator);
+        SEQAN3_RETURN_TYPE_CONSTRAINT(val.erase(val.cbegin(), val.cend()), std::same_as, typename type::iterator);
+    };
 
     SEQAN3_RETURN_TYPE_CONSTRAINT(val.push_back(val.front()), std::same_as, void);
     SEQAN3_RETURN_TYPE_CONSTRAINT(val.push_back(typename type::value_type{}), std::same_as, void);

--- a/test/unit/range/container/container_concept_test.cpp
+++ b/test/unit/range/container/container_concept_test.cpp
@@ -79,6 +79,27 @@ TEST(container, sequence_container_former_travis_bug)
     s.insert(8, " is an example string."s, 0, 14);
     EXPECT_EQ("Exemplar is an example", s);
 
+#if SEQAN3_WORKAROUND_GCC_NO_CXX11_ABI
+    // insert(const_iterator pos, char ch)
+    s.insert(s.begin() + s.find_first_of('n') + 1, ':');
+    EXPECT_EQ("Exemplar is an: example", s);
+
+    // insert(const_iterator pos, size_type count, char ch)
+    s.insert(s.begin() + s.find_first_of(':') + 1, 2, '=');
+    EXPECT_EQ("Exemplar is an:== example", s);
+
+    // insert(const_iterator pos, InputIt first, InputIt last)
+    {
+        std::string seq = " string";
+        s.insert(s.begin() + s.find_last_of('e') + 1,
+            std::begin(seq), std::end(seq));
+        EXPECT_EQ("Exemplar is an:== example string", s);
+    }
+
+    // insert(const_iterator pos, std::initializer_list<char>)
+    s.insert(s.begin() + s.find_first_of('g') + 1, { '.' });
+    EXPECT_EQ("Exemplar is an:== example string.", s);
+#else // ^^^ workaround / no workaround vvv
     // insert(const_iterator pos, char ch)
     s.insert(s.cbegin() + s.find_first_of('n') + 1, ':');
     EXPECT_EQ("Exemplar is an: example", s);
@@ -98,6 +119,7 @@ TEST(container, sequence_container_former_travis_bug)
     // insert(const_iterator pos, std::initializer_list<char>)
     s.insert(s.cbegin() + s.find_first_of('g') + 1, { '.' });
     EXPECT_EQ("Exemplar is an:== example string.", s);
+#endif // SEQAN3_WORKAROUND_GCC_NO_CXX11_ABI
 }
 
 TEST(container, sequence_container)

